src/drivers/intel/cavs/dmic.c

```diff
@@ -153,11 +158,11 @@ static uint64_t dmic_work(void *data)
 	spin_lock(&dai->lock);
 
 	/* Increment gain with logaritmic step.
-	 * Gain is Q2.30 and gain modifier is Q2.14.
+	 * Gain is Q2.30 and gain modifier is Q12.20.
 	 */
 	dmic->startcount++;
-	dmic->gain = Q_MULTSR_32X32((int64_t)dmic->gain,
-				    LOGRAMP_GM, 30, 14, 30);
+	dmic->gain = q_multsr_sat_32x32(dmic->gain, dmic->gain_coef,
+					Q_SHIFT_BITS_32(30, 20, 30));
```

Comment:
just wondering if there would be any merit in passing a pre-computed 'gain_coeff' from topology if we want to reduce the code size and avoid the db/linear conversions in firmware. I feel the control code for DMIC is becoming quite large...

Reply:
The user (device/platform configuration engineer) or kernel or FW can do such conversion. I have most doubts for humans to bother to calculate such parameters. But some user space configuration tool to create topology would solve it.

Eventually there needs to be dB -> lin -> dB conversion in FW so I though this could be a start. The function is generic and e.g. logarithmic ramp in volume control or DRC component could use it too.

For code size I wonder if we could generally optimize for size (-Os) and only some generic C algorithm code like volume, IIR or SRC optimize for speed.

But I'm OK switch to use in topology a raw parameter if you are sure you meant it.

Reply:
@plbossart @lgirdwood @tlauda @lbetlej @mmaka1 @cujomalainey Opinions are welcome for this. This is an overall arch issue, how low/high abstraction level do we want for firmware? Passing decibels is for a fat firmware arch, passing raw numbers is for thin firmware arch (with sometimes tricky looking dependencies in kernel & user space) .

Reply:
From @dgreid

> I'd lean towards passing the coefficient
> Not strongly, but it seems like it would allow for more flexibility in the future

Personally I second that, as more modules get added we might have more types of information being passed and I would be worried about type information getting lost in the increasing number of types being passed.

Reply:
@singalsu pass raw. Let host do conversions.

Reply:
@lgirdwood We have still conflicting comments here. @plbossart  suggested raw parameter into topology while did you mean that the db2lin_fixed() would be executed from sof_link_dmic_load() in topology.c? I'd need to move then some #defines to kernel headers for ramp coefficient computation (initial gain, ramp update rate).

Reply:
@singalsu dB values better in topology as it's written by humans and kcontrol volume already using dB in topology today.  This can be converted to raw during the pre-processing stage. i.e.
1) create new cmd line tool that converts args (dB value) to raw on stdout.
2) invoke this new cmd from M4 dB macro (will replace dB value with raw).

See https://www.gnu.org/software/m4/manual/m4-1.4.15/html_node/Syscmd.html#Syscmd

Reply:
@lgirdwood That gets quite complicated considering that we are talking about quite small amount of generally useful code to be optimized away. 

My rant: It's not just for DMIC. I did it because I felt it would be the right thing to do based on my past experience. Supporting logarithmic volume ramp type needs that. Even dumb & cheap peripherals like codecs do understand decibels and it really simplifies this a lot this even vs. doing the calculation in kernel since it's not just dB to linear conversion.

For this m4 script launch approach to work and remain maintainable I'd need to export to build host user space the the #defines from FW driver that impact ramp time (ms) -> ramp range dB -> ramp step dB -> ramp step linear computation. Can we include directly from e.g. "tools/dmic_ramp_tool" source code the DMIC driver header "../../src/include/sof/dmic.h" to get these constants. Or would they need to be placed to kernel uapi headers?

Reply:
@plbossart wrote:
> reduce the code size and avoid the db/linear conversions in firmware. 

It would seem the cost is small and the expense if not doing in fw is a very complicated FW ABI. If you precompute on host as @lgirdwood described, you expose a lot of new information in the ABI that otherwise could be hidden. You can precompute, but you need much more inputs than the ramp time in millisconds, and all these inputs are now part of your ABI and cannot be changed. If we just have ramp-time-in-msec as the ABI, this seems much more maintainable and futureproof with little fw cost. What am I missing?

Reply:
> For this m4 script launch approach to work and remain maintainable I'd need to export to build host user space the the #defines from FW driver that impact ramp time (ms) -> ramp range dB -> ramp step dB -> ramp step linear computation. Can we include directly from e.g. "tools/dmic_ramp_tool" source code the DMIC driver header "../../src/include/sof/dmic.h" to get these constants. Or would they need to be placed to kernel uapi headers?

@singalsu I don't think users having to manually calculate raw values in order to write topology is maintainable or easy. It's a high barrier to entry if users have to convert dB values into raw values. Can you give an example of all the raw values needed today.

@kv2019i I'm fine with the ABI additions if it makes users life easier, but this depends on how complex/difficult it will be for users to generate rw values from dB values.

Reply:
@lgirdwood but that's the thing, the dB value is not the input, it's the ramp time in milliseconds. This is easy to document, maintain and understand as a FW ABI. If we use "gain_coef" as the ABI, then we also have to have LOGRAMP_START_DB and the CONST_TERM as part of the ABI, and combination results in a more implementation specific and harder to understand&maintain (as an ABI).

Reply:
@cujomalainey  @dgreid I've considered options for not having db2lin() function in firmware but those complicate the system too much. I now even more feel this proposal really is the best way to implement the feature of passing the ramp length as time (milliseconds as integer).

The exp function would be now avoided in code by using instead a linear gain ramp but when I experimented with it, it needed to be at least double length vs. logaritmic. A one second ramp would be likely too much so also I feel this choice was right as well.

Reply:
> Opinions are welcome for this. This is an overall arch issue, how low/high abstraction level do we want for firmware? Passing decibels is for a fat firmware arch, passing raw numbers is for thin firmware arch (with sometimes tricky looking dependencies in kernel & user space) .

> @lgirdwood but that's the thing, the dB value is not the input, it's the ramp time in milliseconds.

@kv2019i @singalsu guys, which is it, are we passing ms or dB or both ?

Reply:
@lgirdwood We pass from topology milliseconds.

Reply:
ok, so where does dB come from as you mentioned earlier? millisecs is fine for users.

Reply:
We have ramp initial gain of g0 = -90 dB (cannot have zero or -inf dB gain as start in log scale, so picked a small value). The -90 dB is defined in a macro. We unmute to  g1 = 0 dB. Our unmute ramp length is t_ramp [ms]. Our ramp update rate is 1000 us, defined in macro.

number of steps = 1000 * t_ramp / update rate

decibel step = (g1 - g0) / number of steps

linear gain coefficient = db2lin_fixed( decibel step)

Reply:
@singalsu ok, who calculates linear gain coefficient and where is this set ? Can you give me an example of what the user has to set and do in topology. My aim here is to keep thing really simple for users so topology would look like
```
GAIN_RAMP(-90, 0, 30, 1000)
```
Where g0 = -90dB, g1 = 0dB, steps = 30 and t_ramp = 1000us. Is this what you have in mind ?

Reply:
@lgirdwood In this PR all calculation is done in FW for those three equations above. Kernel just parses the token and places the value to IPC. The firmware uses hard coded #define macros for initial gain g0 (-90 dB) and ramp update rate (1000 us). The only passed parameter from topology is the length of ramp as milliseconds. The topology part in this PR sets the value to 400 to indicate a 400 ms long gain ramp. 

If we would like to support another ramp like linear or cosine we would add a ramp type token and existing behavior would be the logarithmic ramp type. It's simple and intuitive and why I prefer to to keep this PR as is. Also this work in decibels.c is reusable for new volume component ramp type: logarithmic. 

If we would like to compute the gain coefficient in user space we should have three parameters defined via topology instead: Initial gain (Q1.31, dB -> lin converted), gain update rate (us), gain update coefficient (Q2.30). But if we do that we complicate and freeze the ABI with three parameters those have a low association for  a normal engineer to achieve the task of adjusting the ramp length. The m4 macro could be line GAIN_RAMP(-90, 0, 400, 1000).

But I don't like the idea, too complicated. In my opinion the harm of making more complex than necessary IPC is bigger than the benefit of saving some code lines (24->x). Also the new parameters need sanity check code lines and traced errors plus IPC test cases for parameters combinations. In contrast it was very easy for me to test this PR, I've set in topology the parameter to 10, 400, 1000 and checked from waveform that it happened.

Also if we would need to have other ramp type those IPC parameters for pre-computed log ramp would be useless. The same words would need to be re-purposed for something else pre-calculated for DSP and it would be even more a mess.

Reply:
The PR contains topology changes to set ramp to existing default length of 400 ms. It's as simple as this:

 DAI_CONFIG(DMIC, 0, 6, NoCodec-6,
	   dnl DMIC_CONFIG(driver_version, clk_min, clk_mac, duty_min, duty_max,
	   dnl		   sample_rate, fifo word length, **unmute time**, type,
	   dnl             dai_index, pdm controller config)
	   DMIC_CONFIG(1, 500000, 4800000, 40, 60, 48000,
		DMIC_WORD_LENGTH(s32le), **400**, DMIC, 0,
		PDM_CONFIG(DMIC, 0, DMIC_PDM_CONFIG)))

Reply:
@singalsu ok, thanks - that looks straightforward atm for users. My aim is to keep this simple for users, but I do also want to avoid extra burden on FW for calculating tables/coefficients when this can be done higher up the stack (even if it means more complex IPC). I think we can address both with IPC 2.0

Reply:
OK, I'll take this as accept for now. Right?